### PR TITLE
Add ID command for IMAP

### LINKED ISSUE #5089: Add support for ID command to support 163.com
**Is your feature request related to a problem? Please describe.**
I got rejected receiving mails using imap. I was trying to log in to 163 email server when this happens.

**Describe the solution you'd like**
I was told that the email client app should return its identity to allow connecting to the email server.

The developer of that email server provides the following code:
```
Properties props = new Properties(); 
props.setProperty("mail.store.protocol", "imap"); 
props.setProperty("mail.imap.host", "imap.163.com"); 
props.setProperty("mail.imap.port", "143"); 

HashMap IAM = new HashMap();
//IMAP ID: key, value pair. E.g. name，version，vendor，support-email
IAM.put("name","myname");      
IAM.put("version","1.0.0");      
IAM.put("vendor","myclient");      
IAM.put("support-email","testmail@test.com");  
Session session = Session.getInstance(props); 

IMAPStore store = (IMAPStore) session.getStore("imap");
//user account and password
store.connect("testmail@163.com", "password"); 

store.id(IAM);
```

The developer provides the following link to explain context:
https://javaee.github.io/javamail/docs/api/index.html?com/sun/mail/imap/IMAPStore.html

Could you evaluate if this situation is a valid case to make it more secure/compatible with imap server?

### PR DESCRIPTION
**Pull request by wh201906**
Fixes #5089
This PR is based on the implementation of `useCompression` and PR #5578.
I have tested this PR on Netease mail, which requires the ID extension. It works as expected.

### CODE REVIEW

**Comment by wh201906** [comment]
While this PR is functional, I have some uncertain points about it.
1. The [`loadDefaults()`](https://github.com/thundernest/k-9/blob/9f9062e02ef7e2dd0ace500d07b9c748c93a19f3/app/core/src/main/java/com/fsck/k9/AccountPreferenceSerializer.kt#L569) in [`AccountPreferenceSerializer.kt`](https://github.com/thundernest/k-9/blob/9f9062e02ef7e2dd0ace500d07b9c748c93a19f3/app/core/src/main/java/com/fsck/k9/AccountPreferenceSerializer.kt) contains the default value of [`isIgnoreChatMessages`](https://github.com/thundernest/k-9/blob/9f9062e02ef7e2dd0ace500d07b9c748c93a19f3/app/core/src/main/java/com/fsck/k9/AccountPreferenceSerializer.kt#L581), but I can't find the default value of `useCompression` in this function. Should I add `sendClientID` in it?
2. The [`getBoolean()`](https://github.com/thundernest/k-9/blob/9f9062e02ef7e2dd0ace500d07b9c748c93a19f3/app/ui/legacy/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsDataStore.kt#L24) and [`putBoolean()`](https://github.com/thundernest/k-9/blob/9f9062e02ef7e2dd0ace500d07b9c748c93a19f3/app/ui/legacy/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsDataStore.kt#L48) in [`AccountSettingsDataStore.kt`](https://github.com/thundernest/k-9/blob/9f9062e02ef7e2dd0ace500d07b9c748c93a19f3/app/ui/legacy/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsDataStore.kt) contain the `ignore_chat_messages` ([get](https://github.com/thundernest/k-9/blob/9f9062e02ef7e2dd0ace500d07b9c748c93a19f3/app/ui/legacy/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsDataStore.kt#L43), [set](https://github.com/thundernest/k-9/blob/9f9062e02ef7e2dd0ace500d07b9c748c93a19f3/app/ui/legacy/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsDataStore.kt#L67)), but I can't find `useCompression` or something related in these functions. Should I add `sendClientID`/`send_client_id` in them?
3. [`account_settings.xml`](https://github.com/thundernest/k-9/blob/9f9062e02ef7e2dd0ace500d07b9c748c93a19f3/app/ui/legacy/src/main/res/xml/account_settings.xml) contains the preference of [`ignore_chat_messages`](https://github.com/thundernest/k-9/blob/9f9062e02ef7e2dd0ace500d07b9c748c93a19f3/app/ui/legacy/src/main/res/xml/account_settings.xml#L325-L330), but I can't find preference of `useCompression` or something related in this file. Should I add `sendClientID`/`send_client_id` in it?
4. In this PR, the default value of `sendClientID` is `false`. Should I change it?
5. I didn't write a unit test of this setting. For now I have no idea how this option is tested.
6. How do you think of the name `sendClientID`? Should I change it?
7. The name of K-9 is hard coded in the ID command. should I change it or read it from elsewhere?

File: config/detekt/baseline.xml [hunk 1]
```diff
@@ -183,6 +183,7 @@
...
     <ID>LongParameterList:SaveMessageOperationsTest.kt$SaveMessageOperationsTest$( subject: String? = getSubject(), date: Long = sentDate?.time ?: System.currentTimeMillis(), internalDate: Long = date, downloadState: MessageDownloadState = getDownloadState(), attachmentCount: Int = 0, previewResult: PreviewResult = PreviewResult.none(), textForSearchIndex: String? = null, encryptionType: String? = null, )</ID>
     <ID>LongParameterList:SettingsImportViewModel.kt$SavedAccountState$( val accountIndex: Int, val displayName: String, val accountUuid: String, val selected: Boolean, val importStatus: ImportStatus, val incomingServerName: String?, val outgoingServerName: String?, val incomingServerPasswordNeeded: Boolean, val outgoingServerPasswordNeeded: Boolean, )</ID>
+    <ID>LongParameterList:SimpleImapSettings.kt$SimpleImapSettings$( override val host: String, override val port: Int = 0, override val connectionSecurity: ConnectionSecurity = ConnectionSecurity.NONE, override val authType: AuthType, override val username: String, override val password: String? = null, override val useCompression: Boolean = false, override val sendClientID: Boolean = false, )</ID>
```
[the remark this answers is not in the document]
Comment (wh201906) [reply config/detekt/baseline.xml:186]: I see.

**Comment by wh201906** [comment]
> A handful of providers seem to require the ID command. So it should probably be enabled by default.

> The name of the app ("K-9 Mail") should be supplied via a BuildConfig field. See e.g. 

> Instead of adding a Bool and a String to `ImapSettings`, add `clientId: String?`. That way we can send the ID command if `clientId` is not null, and skip sending the ID command otherwise.

Sorry. I'm not clear about the code for them. Do you mean add a single `String` setting item in the `ImapSettings`? If so, how do the users configure it? IDK if it should be configured by a checkbox or a textfield. Plus, how does this setting item work with the BuildConfig field?
I guess we can add a checkbox in the AccountSetupIncoming.java. If it's checked, write the text from corresponding BuildConfig field to the setting item `clientId`. If not, set the `clientId` to empty. When establishing the connection, 
>  send the ID command if clientId is not null, and skip sending the ID command otherwise.

IDK if this logic is appropriate.

**Comment by cketti** [comment]
> Do you mean add a single `String` setting item in the `ImapSettings`?

Yes.

> If so, how do the users configure it?

They don't. The app will set the value to its app name (read from a BuildConfig field) if the setting to send the client ID is enabled and to `null` otherwise.

**Comment by wh201906** [comment]
By the way, I wonder why not adding a `Boolean` setting item, then read the BuildConfig.clientID when establishing the connection.

**Comment by wh201906** [comment]
I will squash my commits when things are ready.

**Comment by cketti** [comment]
The IMAP code is in a library project that knows nothing about the app using the library. It's also a JVM-only project, that doesn't have access to `BuildConfig` because that's an Android build system thing.

**Comment by wh201906** [comment]
> The IMAP code is in a library project that knows nothing about the app using the library. It's also a JVM-only project, that doesn't have access to BuildConfig because that's an Android build system thing.
I see.

**Comment by cketti** [comment]
Apologies for not being clear enough. `Account` should still only keep track of a boolean value, `isSendClientIdEnabled`.

The code to translate `Account.isSendClientIdEnabled` and the app name into a configuration for the IMAP library should go into `ImapBackendFactory`.

**Comment by wh201906** [comment]
> Apologies for not being clear enough. `Account` should still only keep track of a boolean value, `isSendClientIdEnabled`.
> 
> The code to translate `Account.isSendClientIdEnabled` and the app name into a configuration for the IMAP library should go into `ImapBackendFactory`.

OK. I will refactor my code tomorrow because it's late night(UTC+8).
I think most of the part has been implemented in commit 209a8138418b63d86d487b114f0fd543c435d780. You can check it and give me more suggestions if you want.

**Comment by wh201906** [comment]
If the code is good I'll implement the unit test.

File: mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/RealImapConnection.kt [hunk 2]
```diff
@@ -561,6 +562,12 @@ internal class RealImapConnection(
         }
     }
 
+    private fun sendClientIDIfSupported() {
+        if (hasCapability(Capabilities.ID) && settings.clientId != null) {
+            executeSimpleCommand("ID (" + settings.clientId + ")")
```
Comment (cketti) [inline mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/RealImapConnection.kt:567]: When only handing an app name to the IMAP code, this needs to be changed to:
```suggestion
            val encodedAppName = ImapUtility.encodeString(settings.clientIdAppName)
            executeSimpleCommand("""ID ("name" $encodedAppName)""")
```

**Review by cketti — CHANGES_REQUESTED** [review]

File: mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/ImapSettings.kt [hunk 3]
```diff
@@ -15,6 +15,7 @@ internal interface ImapSettings {
     val password: String?
     val clientCertificateAlias: String?
     val useCompression: Boolean
+    val clientId: String?
```
[the remark this answers is not in the document]
Comment (wh201906) [reply mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/ImapSettings.kt:18]: OK

File: mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/RealImapConnection.kt [hunk 4]
```diff
@@ -561,6 +562,12 @@ internal class RealImapConnection(
         }
     }
 
+    private fun sendClientIDIfSupported() {
```
[the remark this answers is not in the document]
Comment (wh201906) [reply mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/RealImapConnection.kt:565]: OK

File: mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/ImapStoreConfig.kt [hunk 5]
```diff
@@ -4,4 +4,5 @@ interface ImapStoreConfig {
     val logLabel: String
     fun isSubscribedFoldersOnly(): Boolean
     fun useCompression(): Boolean
+    fun clientId(): String?
```
[the remark this answers is not in the document]
Comment (wh201906) [reply mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/ImapStoreConfig.kt:7]: OK

File: mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/ImapSettings.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (wh201906) [reply mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/ImapSettings.kt:18]: The `clientId` also occurs in `createImapStoreConfig()`(ImapBackendFactory.kt) and `StoreImapSettings`(RealImapStore.kt). Should I change them?

[the remark this answers is not in the document]
Comment (cketti) [reply mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/ImapSettings.kt:18]: These sites implement interfaces I asked you to change. So you'll have to change them in order for the code to compile (or the IDE will do it for you if you use the rename feature).

[the remark this answers is not in the document]
Comment (wh201906) [reply mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/ImapSettings.kt:18]: I get it.

**Comment by wh201906** [comment]
Thank you for spending time reviewing this. I have changed the code as you suggested and tested in on my device. It works as 2096b97

**Comment by wh201906** [comment]
> Test cases we should probably have:
> 
>     * ID capability not present → make sure ID command is not sent
>     * ID capability present and setting enabled → make sure ID command is sent
>     * ID capability present and setting disabled → make sure ID command is not sent

Sorry, I don't know what's the server response look like if the ID capability is present. Could you please implement the tests?

**Comment by cketti** [comment]
Thanks :+1:

**Comment by wh201906** [comment]
Thank you so much for guiding me and reviewing this!

**Pull request merged**